torch/nn/_functions/thnn/pooling.py

```diff
@@ -364,6 +364,145 @@ def backward(self, grad_output):
+            self.mark_non_differentiable(indices)
+            return output, indices
+        else:
+            self.save_for_backward(input)
+            self.indices = indices
+            return output
+
+    def backward(self, grad_output, _indices_grad=None):
+        if self.return_indices:
+            input, indices = self.saved_tensors
+        else:
+            input, = self.saved_tensors
+            indices = self.indices
+
+        input2d = input.unsqueeze(2)
+        indices2d = indices.unsqueeze(2)
+        grad_output2d = grad_output.unsqueeze(2)
+        grad_input = grad_output2d.new()
+        backend = type2backend[type(input)]
+        backend.SpatialAdaptiveMaxPooling_updateGradInput(backend.library_state,
+                                                          input2d, grad_output2d, grad_input, indices2d)
+        grad_input = grad_input.squeeze(2)
+        return grad_input
+
+
+class AdaptiveMaxPool2d(Function):
+
+    def __init__(self, output_size, return_indices=False):
+        self.output_size = _pair(output_size)
+        self.return_indices = return_indices
+
+    def forward(self, input):
+        backend = type2backend[type(input)]
+        indices, output = input.new().long(), input.new()
+        backend.SpatialAdaptiveMaxPooling_updateOutput(backend.library_state,
+                                                       input, output, indices,
+                                                       self.output_size[1], self.output_size[0])
+        if self.return_indices:
+            self.save_for_backward(input, indices)
+            self.mark_non_differentiable(indices)
+            return output, indices
+        else:
+            self.save_for_backward(input)
+            self.indices = indices
+            return output
+
+    def backward(self, grad_output, _indices_grad=None):
+        if self.return_indices:
+            input, indices = self.saved_tensors
+        else:
+            input, = self.saved_tensors
+            indices = self.indices
+        grad_input = grad_output.new()
+        backend = type2backend[type(input)]
+        backend.SpatialAdaptiveMaxPooling_updateGradInput(backend.library_state,
+                                                          input, grad_output, grad_input, indices)
+        return grad_input
+
+
+class AdaptiveAvgPool1d(Function):
+
+    def __init__(self, output_size):
+        self.output_size = _single(output_size)
+
+    def forward(self, input):
+        if input.dim() != 3:
+            raise ValueError('expected 3D input (got {}D input)'
+                             .format(input.dim()))
+
+        input2d = input.unsqueeze(2)    # size = N*C*1*L
+        backend = type2backend[type(input)]
+        output = input2d.new()
+        self.save_for_backward(input)
+        backend.SpatialAdaptiveAveragePooling_updateOutput(
+            backend.library_state,
+            input2d, output,
+            self.output_size[1], self.output_size[0])
```

Comment:
Shouldn't one of the sizes be a 1 constant here? How could that even pass the tests?